Revision 0173-Read-Generic-Network-Signal-data.md

atiwari9 opened the pull request:
## Introduction
This is a revision to an accepted, but not yet implemented proposal. The suggested revisions came out of discussions during implementation of the original proposal. These changes should not cause a major version change.

## Motivation

We need to have a clear way of versioning vehicle data schema file and OEM mapping table so that data exchange between module and cloud is minimal. Along with these changes, we are also going to provide some clarifications for implementation detail questions. We will be addressing following points:

* To include versioning for Vehicle Data Schema
* To specify versioning and endpoint for OEM mapping table and to provide HMI a way to read OEM mapping table version
* To use snake case for _VehicleDataItems_ key
* To define SDL request structure for Get/Subscribe vehicle data	
* To define use cases for vehicle data schema update


## Proposed solution

### To include versioning for Vehicle Data Schema
Since vehicle data schema will be downloaded as part of policy update, we should not send this data with every update when there is no change in schema. One solution could be that we include the schema in sdl_snapshot when requesting policy update, but that again does not solve the issue as snapshot gets large this way.

#### Solution:
Solution to this is to include a schema versioning. This version information will be sent along with PTU when schema update is sent to module in PTU and SDL will persist this value in local policy DB. sdl_snapshot will contain this schema version information when requesting the policy update. Policy server would then compare the schema version received from cloud with what is available on server. Based on this comparison, vehicle data schema update would be sent in policy update response. e.g.

```
"vehicle_data_schema_version":"00.00.01",
```

### To specify versioning and endpoint for OEM mapping table and to provide HMI a way to read OEM mapping table version
Similarly, OEM mapping table needs to be downloaded by HMI using _SystemRequest_ type _OEM_SPECIFIC_. HMI has no way to know if this version is updated on cloud, so either HMI keeps downloading it at set interval which is too inefficient or we can introduce OEM mapping table version as well.

#### Solution:
OEM mapping table version would be included in PTU, which SDL will persist in local policy DB. HMI needs to be able to read this value on demand so that it can control if and when to download the OEM mapping table. HMI will utilize the _endpoint_ for OEM mapping table file to download this file using _SystemRequest_ type _OEM_SPECIFIC_. OEM mapping table file would have _endpoint_ key (_service_) as  _oem_mapping_file_.

SDL still needs to provide an API for HMI to read this value, we have two options for that:

1. Reuse existing API _GetUserFriendlyMessage_ with _messageCodes_ as _oem_mapping_version_ and any language.

2. Add new API to read OEM mapping file version, we can make it generic so that it can also be used in future to read some other configuration params from policy. e.g. below new API can be used to read OEM mapping table version with _identifier_ _oem_mapping_version_

```
<function name="GetPolicyConfigurationData" messagetype="request" scope="internal">
	<description>Request from HMI to SDL to get policy configuration data (i.e. OEM mapping table file version etc.) from Policy Table.</description>
	<param name="identifier" type="String" array="true" minsize="1" maxsize="100" maxlength="500" mandatory="true">
		<description>Id of configuration data to be received according to Policy Table i.e. oem_mapping_version etc.</description>
	</param>  
</function>
<function name="GetPolicyConfigurationData" messagetype="response">
	<param name="identifierValue" type="String" array="true" minsize="1" maxsize="100" mandatory="false">
		<description>If no message was found in PT for specified message code and for HMI current or specified language, this parameter will be omitted.</description>
	</param>
</function>
```

We prefer second option, as it is purpose build and can be extended to be used for future configuration items as well.

### To use snake case for _VehicleDataItems_ key
Currently vehicle data schema object is camel case (_VehicleDataItems_) in accepted proposal and rest of policy table items are in snake case.

#### Solution:
Change the case to snake case as: _vehicle_data_items_

With above three changes, following is how the new structure would look like:

```
"vehicle_data":{
	"oem_mapping_version":"00.00.01",
	"vehicle_data_schema_version":"00.00.01",
	"vehicle_data_items": [
			{
				"name": "rpm",
				"type": "Integer",
				"key": "OEM_REF_RPM",
				"array": false,
				"mandatory": false,
				"params": [

				]
			},
			...
		]
}
```

Along with above structure, there are some rules which need to be followed:
* These two variables (_oem_mapping_version_ & _vehicle_data_schema_version_) would be fed by Policy server to module during PTU. _oem_mapping_version_ would always be included in a PTU, unlike _vehicle_data_items_ and _vehicle_data_schema_version_, which would only be present if there is a change/update in vehicle data items. 
* SDL would need to include _vehicle_data_schema_version_ in sdl_snapshot while requesting the policy update. SDL  policy server would this to decide whether _vehicle_data_items_ schema needs to be pushed in PTU response.
  * _vehicle_data_schema_version_ would only be included in _vehicle_data_ only if _vehicle_data_items_ schema is included.
* SDL would need to provide a way for HMI to read _oem_mapping_version_ so that HMI can decided whether to request OEM mapping file download using the _endpoint_ mentioned in PTU for the OEM table mapping file, using _SystemRequest_ type _OEM_SPECIFIC_


### To define SDL request structure for Get/Subscribe vehicle data

In case of VehicleDataItems defined in API, SDL should send to HMI 'dataItemName : true' of item according to HMI_API.
HMI should work as is to provide response on such request. 

In case of VehicleData Items from schema, SDL should provide hierarchical structure description of vehicle data item.

We propose to send schema vehicle data Items with following structure : 
```
{
	"dataItemKey" : // Struct with parameters 
}
```

Here is an example for "headLampStatus"

#### HMI request response approach for vehicle data item present in API:

##### Request: 
```json
{
	"headLampStatus" : "true"	
}
```

##### Response: 
```json
{
	"headLampStatus" : {
		"lowBeamsOn" : true,
		"highBeamsOn" : true,
		"ambientLightSensorStatus" : "TWILIGHT_1"
	}
}
```

#### Consider headLampStatus sent as schema vehicle data item : 


##### Request:
```json
{
	"HEAD_LAMP_STATUS" : {
		"HEAD_LAMP_STATUS_LEFT" : "true",
		"HEAD_LAMP_STATUS_RIGHT" : "true",
		"AMBIENT_LIGHT_SENSOR_STATUS" : "true"
	}
}
```

##### Response:
```json
{
	"HEAD_LAMP_STATUS" : {
		"HEAD_LAMP_STATUS_LEFT" : true,
		"HEAD_LAMP_STATUS_RIGHT" : true,
		"AMBIENT_LIGHT_SENSOR_STATUS" : "TWILIGHT_1"
	}
}
```

Note that in request SDL provided list of nested "keys" of structure. 

#### Example of structure with 3 nested levels.

##### Request: 
```json
{
	"KEY_LEVEL_1" : {
		"KEY_LEVEL_2" : "true",
		"KEY2_LEVEL_2" : {
			"KEY2_LEVEL_3" : "true"
		}
	}
}
```

##### Request: 
```json
{
	"KEY_LEVEL_1" : {
		"KEY_LEVEL_2" : "value of key level 2",
		"KEY2_LEVEL_2" : {
			"KEY2_LEVEL_3" : "value of key level 3"
		}
	}
}
```
#### Backward compatibility 

Such approach is backward compatible. This also provides ability to send vehicle data items present in API and vehicle data items present in schema in one request.

#### Mix example 

Consider **gps** as API vehicle data Item, and **HEAD_LAMP_STATUS** as vehicle data item from schema: 

##### Request:
```json
{
	"gps" : "true", 
	"HEAD_LAMP_STATUS" : {
		"HEAD_LAMP_STATUS_LEFT" : "true",
		"HEAD_LAMP_STATUS_RIGHT" : "true",
		"AMBIENT_LIGHT_SENSOR_STATUS" : "true"
	}
}
```

##### Response: 

```json
{
	"gps" : {
		"longtitudeDegrees" : 0.0,
		"latitudeDegrees" : 0.0
	},
	"HEAD_LAMP_STATUS" : {
		"HEAD_LAMP_STATUS_LEFT" : true,
		"HEAD_LAMP_STATUS_RIGHT" : true,
		"AMBIENT_LIGHT_SENSOR_STATUS" : "TWILIGHT_1"
	}
}
```

#### Rules for SDL<->HMI structure for requests and responses

- HMI should consider parameter in GVD request as name in case if it exist in list HMI_API.xml GetVehicleData of parameters.
- HMI should consider parameter in GVD request as key if it is not in list of HMI_API.xml GetVehicleData of parameters.
- _name_ is always boolean (according to API)
- _key_ in GVD request may be either boolean or struct. Struct may contain either booleans or structs. 
- Response type for vehicle data items from schema would be _VehicleDataResult_
- If HMI response to SDL with wrong item schema for GVD, SDL will not process this response and respond to mobile with GENERIC_ERROR
- If HMI send to SDL _OnVehicleData_ Notification with wrong item schema, SDL will ignore this notification.
- In case of UnsubscribeVehicleData, SDL will send only root level **key**.

### To define use cases for vehicle data schema update
This section defines SDL behavior in case _vehicle_data_items_ schema is not present in PTU or if data in schema does not match with API/exiting schema. So if in a PTU:
* _vehicle_data_items_ key does not exist; do not update _vehicle_data_items_ local schema/items.
* _vehicle_data_items_ exists; replace all local schema vehicle data items with new ones except the standard ones.
* _vehicle_data_items_ exists but is empty; remove all local schema vehicle data items except the standard ones.
* _vehicle_data_items_ has item with same name as one of standard vehicle data items, ignore those data items.

** Standard vehicle data items are the ones which are, either present in API or are provided in base schema with SDL core release. Standard vehicle data items are immutable.


## Proxy side changes
None


## SDL_Server side changes
* SDL server would need to change the schema structure as defined above and add versioning variables:
```
"oem_mapping_version":"00.00.01",
"vehicle_data_schema_version":"00.00.01",

```
* SDL Server would need to compare the version received from sdl_snapshot before serving schema update.

## Impact on existing code
* Since this is a revision to proposal currently being implemented, any change would be absorbed in the same implementation. Having said that, there would be minor changes around:

  * New API is needed to provide HMI a way to read OEM mapping table version
  * SDL would need to persist OEM and schema versions in policy DB
  * SDL would need to add schema version to sdl_snapshot


## Alternatives considered
* We can avoid using versioning information if:
  * SDL adds schema to sdl_snapshot
  * Based on changes for schema, downloads OEM mapping table using _SystemRequest_ type _OEM_SPECIFIC_ with OEM mapping table _endpoint_
    * Implements a retry strategy i.e. 5 times every 60s for unsuccessful OEM file download every ignition cycle

atiwari9 commented:
@theresalech - Ready for review

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -62,101 +62,154 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
+* _vehicle_data_items_ exists but is empty; remove all local schema vehicle data items except the standard ones.
+* _vehicle_data_items_ has item with same name as one of standard vehicle data items, ignore those data items.
+
+*Standard vehicle data items are the ones which are, either present in API or are provided in base schema with SDL core release. Standard vehicle data items are immutable.
```
```suggestion
*Standard vehicle data items are the ones which are either present in API, or are provided in base schema with SDL core release. Standard vehicle data items are immutable.*
```

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -62,101 +62,154 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
 
+#### Use cases for vehicle data schema update:
+This section defines SDL behavior in case _vehicle_data_items_ schema is not present in PTU or if data in schema does not match with API/exiting schema. So if in a PTU:
+* _vehicle_data_items_ key does not exist; do not update _vehicle_data_items_ local schema/items.
```
```suggestion
* _vehicle_data_items_ key does not exist, do not update _vehicle_data_items_ local schema/items.
```

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -62,101 +62,154 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
+#### Use cases for vehicle data schema update:
+This section defines SDL behavior in case _vehicle_data_items_ schema is not present in PTU or if data in schema does not match with API/exiting schema. So if in a PTU:
+* _vehicle_data_items_ key does not exist; do not update _vehicle_data_items_ local schema/items.
+* _vehicle_data_items_ exists; replace all local schema vehicle data items with new ones except the standard ones.
```
```suggestion
* _vehicle_data_items_ exists, replace all local schema vehicle data items with new ones except the standard ones.
```

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -62,101 +62,154 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
+This section defines SDL behavior in case _vehicle_data_items_ schema is not present in PTU or if data in schema does not match with API/exiting schema. So if in a PTU:
+* _vehicle_data_items_ key does not exist; do not update _vehicle_data_items_ local schema/items.
+* _vehicle_data_items_ exists; replace all local schema vehicle data items with new ones except the standard ones.
+* _vehicle_data_items_ exists but is empty; remove all local schema vehicle data items except the standard ones.
```
```suggestion
* _vehicle_data_items_ exists but is empty, remove all local schema vehicle data items except the standard ones.
```

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -62,101 +62,154 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
+
+#### Versioning and endpoint for OEM mapping table and to provide HMI a way to read OEM mapping table version
+
+OEM mapping table version would be included in PTU, which SDL will persist in local policy DB. HMI needs to be able to read this value on demand so that it can control if and when to download the OEM mapping table. HMI will utilize the _endpoint_ for OEM mapping table file to download this file using _SystemRequest_ type _OEM_SPECIFIC_. OEM mapping table file would have _endpoint_ key (_service_) ...
```
```suggestion
OEM mapping table version would be included in PTU, which SDL will persist in local policy DB. HMI needs to be able to read this value on demand so that it can control if and when to download the OEM mapping table. HMI will utilize the _endpoint_ for OEM mapping table file to download this file using _SystemRequest_ type _OEM_SPECIFIC_. OEM mapping table file would have _endpoint_ key (_service_) as _oem_mapping_file_.
```

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -62,101 +62,154 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
+
+OEM mapping table version would be included in PTU, which SDL will persist in local policy DB. HMI needs to be able to read this value on demand so that it can control if and when to download the OEM mapping table. HMI will utilize the _endpoint_ for OEM mapping table file to download this file using _SystemRequest_ type _OEM_SPECIFIC_. OEM mapping table file would have _endpoint_ key (_service_) ...
+
+SDL still needs to provide an API for HMI to read this value, we can do so by adding new API to read OEM mapping file version, we can make it generic so that it can also be used in future to read some other configuration params from policy. e.g. following new API can be used to read OEM mapping table version with _identifier_ _oem_mapping_version_
```
```suggestion
SDL still needs to provide an API for HMI to read this value. We can do so by adding a new API to read OEM mapping file version; we can make it generic so that it can also be used in the future to read some other configuration params from policy. e.g. following new API can be used to read OEM mapping table version with _identifier_ _oem_mapping_version_
```

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -62,101 +62,154 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
+This API can also be extended to be used for future configuration items as well.
+
+##### Rules for OEM mapping table and schema versioning:
+* These two variables (_oem_mapping_version_ & _vehicle_data_schema_version_) would be fed by Policy server to module during PTU. _oem_mapping_version_ would always be included in a PTU, unlike _vehicle_data_items_ and _vehicle_data_schema_version_, which would only be present if there is a change/update in vehicle data items. 
```
```suggestion
* These two variables (_oem_mapping_version_ & _vehicle_data_schema_version_) would be fed by policy server to module during PTU. _oem_mapping_version_ would always be included in a PTU, unlike _vehicle_data_items_ and _vehicle_data_schema_version_, which would only be present if there is a change/update in vehicle data items. 
```

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -62,101 +62,154 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
+* SDL would need to include _vehicle_data_schema_version_ in sdl_snapshot while requesting the policy update. SDL  policy server would use this to decide whether _vehicle_data_items_ schema needs to be pushed in PTU response.
+  * _vehicle_data_schema_version_ would only be included in _vehicle_data_ only if _vehicle_data_items_ schema is included.
+* SDL should skip the _vehicle_data_Items_ update in case _vehicle_data_schema_version_ is not included along with it.
+* SDL would need to provide a way for HMI to read _oem_mapping_version_ so that HMI can decided whether to request OEM mapping file download using the _endpoint_ mentioned in PTU for the OEM table mapping file, using _SystemRequest_ type _OEM_SPECIFIC_
```
```suggestion
* SDL would need to provide a way for HMI to read _oem_mapping_version_ so that HMI can decide whether or not to request OEM mapping file download using the _endpoint_ mentioned in PTU for the OEM table mapping file, using _SystemRequest_ type _OEM_SPECIFIC_
```

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -357,6 +410,142 @@ Sample response from module:
+
+In case of VehicleData Items from schema, SDL should provide hierarchical structure description of vehicle data item.
+
+We propose to send schema vehicle data Items with following structure : 
```
```suggestion
We propose to send schema VehicleData Items with following structure : 
```

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -357,6 +410,142 @@ Sample response from module:
+
+#### Rules for SDL<->HMI structure for requests and responses
+
+- HMI should consider parameter in GVD request as name in case if it exist in list HMI_API.xml GetVehicleData of parameters.
```
```suggestion
- HMI should consider parameter in GVD request as name in case if it exists in list HMI_API.xml GetVehicleData of parameters.
```

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -357,6 +410,142 @@ Sample response from module:
+- _name_ is always boolean (according to API)
+- _key_ in GVD request may be either boolean or struct. Struct may contain either booleans or structs. 
+- Response type for vehicle data items from schema would be _VehicleDataResult_
+- If HMI response to SDL with wrong item schema for GVD, SDL will not process this response and respond to mobile with GENERIC_ERROR
```
```suggestion
- If HMI responds to SDL with wrong item schema for GVD, SDL will not process this response and respond to mobile with GENERIC_ERROR
```

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -357,6 +410,142 @@ Sample response from module:
+- _key_ in GVD request may be either boolean or struct. Struct may contain either booleans or structs. 
+- Response type for vehicle data items from schema would be _VehicleDataResult_
+- If HMI response to SDL with wrong item schema for GVD, SDL will not process this response and respond to mobile with GENERIC_ERROR
+- If HMI send to SDL _OnVehicleData_ Notification with wrong item schema, SDL will ignore this notification.
```
```suggestion
- If HMI sends to SDL _OnVehicleData_ Notification with wrong item schema, SDL will ignore this notification.
```

atiwari9 commented:
@theresalech  - I have updated the proposal with changes you had recommended in review comments.

atiwari9 commented:
@theresalech  - I have updated the PR as per summary comments [this](https://github.com/smartdevicelink/sdl_evolution/issues/768#issuecomment-510161449) and [this](https://github.com/smartdevicelink/sdl_evolution/issues/768#issuecomment-509824528)

I am still awaiting response on point 14, so this update does not reflect any discussion we have had on that point.

atiwari9 commented:
@theresalech  - Ready for review

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -16,9 +16,13 @@ This proposal is to make vehicle data APIs (GetVehicleData/SubscribeVehicleData/
 ## Proposed solution
 
+#### Definitions for terms used in proposal:
+* OEM Network Mapping: This file referse to key value pairs that the OEM module will use to retrieve vehicle data off its network
```
>This file referse to key

Please update to "This file refers to key..."

atiwari9 replied:
Done

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -16,9 +16,13 @@ This proposal is to make vehicle data APIs (GetVehicleData/SubscribeVehicleData/
 
+#### Definitions for terms used in proposal:
+* OEM Network Mapping: This file referse to key value pairs that the OEM module will use to retrieve vehicle data off its network
+* Vehicle Data Schema: This is a JSON object that contains vehicle data defintions for both requests and responses that can be sent using
```
>vehicle data defintions

Please update to "vehicle data definitions"

atiwari9 replied:
Done

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -63,53 +71,69 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
-* _vehicle_data_items_ has item with same name as one of standard vehicle data items, ignore those data items.
-
-**Standard vehicle data items are the ones which are either present in API, or are provided in base schema with SDL core release. Standard vehicle data items are immutable.**
+This section defines SDL core behavior in case `schema_items` is not present in PTU or if data in schema does not match with API/exiting schema. So if in a PTU:
```
>API/exiting schema

Should this be "API/existing schema"?

atiwari9 replied:
Changed to RPC spec

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -63,53 +71,69 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
-	<param name="identifierValue" type="String" array="true" minsize="1" maxsize="100" mandatory="false">
-		<description>If no message was found in PT for specified message code and for HMI current or specified language, this parameter will be omitted.</description>
+	<param name="value" type="String" array="true" maxlength="1000" minsize="1" maxsize="100" mandatory="false">
+		<description>Value of requested property from polocyType in PT. If no value is found in PT for specified policyType and property, this parameter will be omitted.</description>
```
>polocyType

Please update to "policyType"

atiwari9 replied:
Done

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -63,53 +71,69 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
-		<description>Id of configuration data to be received according to Policy Table i.e. oem_mapping_version etc.</description>
+	<description>Request from HMI to SDL core to get policy configuration data (i.e. OEM Network Mapping table file version etc.) from Policy Table.</description>
+	<param name="policyType" type="String" maxlength="1000" minlength="1" mandatory="true">
+		<description>Name of the Struct from where configuration data to be received according to Policy Table types i.e. module_config etc.</description>
```
>Name of the Struct from where configuration data to be received according to Policy Table types i.e. module_config etc.

Please update to "Name of the Struct where configuration data is located in Policy Table, i.e. module_config etc."

atiwari9 replied:
Done

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -63,53 +71,69 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
+		<description>Name of the Struct from where configuration data to be received according to Policy Table types i.e. module_config etc.</description>
+	</param>  
+	<param name="property" type="String" maxlength="1000" minlength="1" mandatory="true">
+		<description>Name of the variable for which the configuration data to be received according to Policy Table types i.e. vehicle_year etc.</description>
```
>Name of the variable for which the configuration data to be received according to Policy Table types i.e. vehicle_year etc.

Please update to "Name of the property located within the policyType Struct, i.e. vehicle_year etc."

atiwari9 replied:
Done

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -63,53 +71,69 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
 
+* HMI sends _GetPolicyConfigurationData_ request to core to read `oem_mapping_version` from policy
+* Core responds with available `oem_mapping_version`
+* HMI compares the `oem_mapping_version` with local copy and dedides if OEM Network Mapping table needs to be updated
```
> dedides

Please update to "decides"

atiwari9 replied:
N/A, outdated

theresalech commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -63,53 +71,69 @@ A base JSON schema file for SDL core and proxy will be kept locally and an addit
+Here is the flow diagram SDL server and core need to follow:
+![Schema_Version_Update_flow](../assets/proposals/0173-Read-Generic-Network-Signal-data/Schema_Version_Update_flow.png)
+
+* `version mismatch` decision box simply checks if `schema_version` recevied in snapshot equals to `schema_version` available at server side.
```
> recevied

Please update to "received"

atiwari9 replied:
Done

atiwari9 commented:
@theresalech  updated the PR as per review comments. Also, i tracked changes made to this PR in below rough write up for reference.

```



Author to Update
Done 1. Author to ensure proposal mentions SHAID and SDL Server as affected platforms.
Done 2. Author to add a requestSubType mention to proposal.
Done 3. Author to update GetPolicyConfiguration RPC in the proposal.
Done 4. Author to update proposal with SystemRequest flow chart and description of the process.
Done 8. Author to update proposal to simplify Core behavior when a vehicle data schema item update occurs (truncate-and-insert rather than search-delete-and-insert).
Done 9. Author to include Policy Table endpoints sample JSON in proposal to clarify endpoint changes and use recommended naming.
Done 10. Author to update proposal to clarify that vehicle data schema version checking is a simple equality check (not semantic versioning).
Done 12. See 16.
Done 13. Author to update proposal to indicate an additional field be added to the “Module Config” tab “Endpoints” section of SDL Server to allow an OEM to set custom_vehicle_data_mapping_url.
Done 15. Author to update proposal to remove redundant Policy Table key names/prefixes.
Done 16. Author to update proposal to use SDL specific and consistent, defined terminology throughout proposal. This will include the addition of flow charts supplied in the comments.
Done 17. Author should revise proposal to state proxy library changes will include only those represented in the RPC spec; remove reference to a “local schema” for the proxy library.
Done 18. Author to consider including this information as scoped work into the proposal
Done 19. Author to update example to be correct.
Done as per most recent aggrement 20. Author to updated method names to be conforming so that feature is easily traceable to all projects. Based on URL rename, suggest to go with customVehicleData.

Done 9. Will use 0x0A as endpoint
Done, update PR 14. Further discussion on the review issue will be required to confirm next steps. The review issue will be left unlocked so that this conversation can take place while the author is revising the proposal in a separate PR.
Added a comment 17. Standard RPCs found in the RPC spec will need to be included in the vehicle data schema used by Core, so that if the RPC Spec is updated with a new vehicle data item, the old schema with OEM-specific vehicle data items can be updated to use the new version of the defintions of those vehicle data items as added in the RPC Spec. This will allow Core to dynamically add new vehicle data items as they become available for all apps using the accepted, standard vehicle data items. Therefore, the schema will be part of preloaded and updated schema Core uses, however the mobile side will not need this schema as the libraries will always implement the latest agreed upon vehicle data items.
Done , Ommitted 21. Further discussion on the review issue will be required, once the Project Maintainer has confirmed Core’s response when a specific endpoint is not provided.
```

atiwari9 commented:
@theresalech  Updated the proposal with changes as per [this](https://github.com/smartdevicelink/sdl_evolution/issues/768#issuecomment-514214361) comment.

atiwari9 commented on proposals/0173-Read-Generic-Network-Signal-data.md:
```diff
@@ -152,21 +152,21 @@ Example for OEM Network Mapping table file `endpoint` and `version` in PTU, plea
-	<param name="value" type="String" array="true" maxlength="1000" minsize="1" maxsize="100" mandatory="false">
-		<description>Value of requested property from policyType in PT. If no value is found in PT for specified policyType and property, this parameter will be omitted.</description>
+	<param name="value" type="String" array="true" minsize="1" maxsize="100" mandatory="false">
+		<description>Value of requested property from polocyType in PT. If no value is found in PT for specified policyType and property, core should respond to HMI with DATA_NOT_AVAILABLE. maxlength of the value parameter is system dependent.</description>
```
Ahh, copy/paste. Corrected it now.

atiwari9 commented:
@theresalech  - Updated!!
